Use fewer connections in the multi-version client

sfc-gh-ajbeamon opened the pull request:
This updates the multi-version client to monitor the protocol version of the cluster and use only the appropriately-versioned client to connect. For clusters older than 7.0, it will also be necessary to monitor the cluster version from the primary client.

As a result, this should eliminate the impact of incompatible clients on the cluster.

- Passed 10K correctness tests
- Passed fdb_c_unit_tests
- Passed 10K binding tester runs (when combined with fix in #4675)
- Passed multithreaded_client.py tests

The following real-world tests were performed to verify that the client functioned and that we had the expected number of connections:

- 5.0 external, cluster at 5.0
- 5.1 and 6.3 external, cluster at 5.1
- 6.1, 6.2, and 6.3 external; cluster at 6.2
- 6.2 and 6.3 external; cluster at 6.3
- 6.2 and 6.3 external; cluster at 6.3 (local client disabled)
- 6.2, 6.3, and 7.0 external; cluster at 7.0
- 6.2, 6.3, and 7.0 external; cluster at 7.0 (local client disabled)
- 5.0 and 5.1 external, cluster upgraded from 5.0 to 5.1
- 5.1, 5.2, and 6.3 external; cluster upgraded from 5.1 to 5.2
- 5.1 and 6.3 external; cluster upgraded from 5.1 to 6.3
- 6.2 and 6.3 external; cluster upgraded from 6.2 to 6.3
- 6.2 and 6.3 external; cluster upgraded from 6.2 to 7.0
- 6.2 and 6.3 external, cluster upgraded from 6.3 to 7.0
- 6.2, 6.3, and 7.0 external, cluster upgraded from 6.3 to 7.0 (local client disabled)
- 6.3 external, cluster upgraded from 6.2 to 6.3 (no connection before upgrade)
- 6.2 external, cluster upgraded from 6.2 to 6.3 (no connection after upgrade)
- 6.2 external, cluster upgraded from 6.2 to 6.3 to 7.0 (no connection in the middle)
- 6.2 and 7.0 external, cluster upgraded from 6.2 to 6.3 to 7.0 (local client disabled, no connection in the middle)
- 5.0 and 6.2 external, cluster "downgraded" from 6.2 to 5.0
- 5.1 and 6.2 external, cluster "downgraded" from 6.2 to 5.1
- 5.0 external, cluster "downgraded" from 7.0 to 5.0
- 5.1 external, cluster "downgraded" from 7.0 to 5.1
- 6.3 external, cluster "downgraded" from 7.0 to 6.3

Note that the downgrade from 7.0 to 5.1 has spotty success (it sometimes works and sometimes doesn't), but best I can tell this is a behavior of the 5.1 client and not specifically a problem of the new implementation. Possibly aspects of the new client make the behavior in 5.1 more likely to trigger.

# Code-Reviewer Section

The general guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [ ] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [ ] Every function/class/actor that was touched is reasonably well documented.

dongxinEric commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -4872,48 +4886,97 @@ Future<Standalone<StringRef>> Transaction::getVersionstamp() {
-	state ClientCoordinators coord(f);
+// Gets the protocol version reported by a coordinator via the protocol info interface
+ACTOR Future<ProtocolVersion> getCoordinatorProtocol(NetworkAddressList coordinatorAddresses) {
+	RequestStream<ProtocolInfoRequest> requestStream{ Endpoint{ { coordinatorAddresses }, WLTOKEN_PROTOCOL_INFO } };
```
Should this endpoint be included into the `ClientLeaderRegInterface`?

sfc-gh-ajbeamon replied:
It's a good question. I think probably it's out of scope of this PR, but it's worth thinking about whether there's a better way to expose this endpoint.

Possibly we could do something like your suggestion, though that may not be the most natural place for it given this endpoint is available on all workers. You could do something like the `WorkerInterface` maybe, though in that case I'd have to see if it's awkward to get a `WorkerInterface` for a process that we haven't connected to and may not be compatible with.

dongxinEric replied:
It's used in `getCoordinatorProtocol` and I thought hmm that endpoint should be part of the interface that a client process uses to interact with Coordinators, which in this case is `ClientLeaderRegInterface`. Why isn't it the most natural place?

sfc-gh-ajbeamon replied:
Every worker exposes this interface, and we just happen to be getting it from a coordinator in this case.

dongxinEric replied:
Ah, I see. It is indeed tricky to find a good place for it.

dongxinEric commented:
Really appreciate the comments!!

sfc-gh-ajbeamon commented:
Marking as draft until I resolve an issue I recently discovered.

sfc-gh-ajbeamon commented:
I fixed an issue when using older API versions that could result in crashes. I've updated the tested scenarios to include some older FDB versions as well.

sfc-gh-anoyes commented on fdbclient/MultiVersionTransaction.h:
```diff
@@ -437,82 +437,114 @@ class MultiVersionDatabase final : public IDatabase, ThreadSafeReferenceCounted<
+		ThreadFuture<Void> protocolVersionMonitor;
+		std::list<LegacyVersionMonitor> legacyVersionMonitors;
+		Optional<ProtocolVersion> dbProtocolVersion;
+		std::map<ProtocolVersion, Reference<ClientInfo>> clients;
```
Mention that this map is keyed by normalized protocol version

sfc-gh-anoyes commented on fdbclient/MultiVersionTransaction.h:
```diff
@@ -437,82 +437,114 @@ class MultiVersionDatabase final : public IDatabase, ThreadSafeReferenceCounted<
+	// A struct that enables monitoring whether the cluster is running an old version (<= 5.0) that doesn't support
+	// connect packet monitoring.
+	struct LegacyVersionMonitor {
+		LegacyVersionMonitor(Reference<ClientInfo> client) : client(client), monitorRunning(false) {}
```
Nit: pass by const reference

sfc-gh-anoyes commented on fdbclient/MultiVersionTransaction.actor.cpp:
```diff
@@ -963,189 +970,249 @@ double MultiVersionDatabase::getMainThreadBusyness() {
-// Returns the protocol version reported by a quorum of coordinators
+// Returns the protocol version reported by the coordinator this client is connected to
 // If an expected version is given, the future won't return until the protocol version is different than expected
 ThreadFuture<ProtocolVersion> MultiVersionDatabase::getServerProtocol(Optional<ProtocolVersion> expectedVersion) {
```
Can this future become ready if the server is older than 6.2? Is that documented anywhere? I think we're technically exposing this through the c api so maybe the documentation could go there.

sfc-gh-ajbeamon replied:
Yes, but not if it's older than 5.1. I can add a note about that.

sfc-gh-anoyes commented on fdbclient/MultiVersionTransaction.actor.cpp:
```diff
@@ -963,189 +970,249 @@ double MultiVersionDatabase::getMainThreadBusyness() {
+
+MultiVersionDatabase::DatabaseState::DatabaseState(std::string clusterFilePath, Reference<IDatabase> versionMonitorDb)
+  : clusterFilePath(clusterFilePath), versionMonitorDb(versionMonitorDb),
+    dbVar(new ThreadSafeAsyncVar<Reference<IDatabase>>(Reference<IDatabase>(NULL))) {}
```
```suggestion
    dbVar(new ThreadSafeAsyncVar<Reference<IDatabase>>(Reference<IDatabase>(nullptr))) {}
```

sfc-gh-anoyes commented on fdbclient/MultiVersionTransaction.actor.cpp:
```diff
@@ -963,189 +970,249 @@ double MultiVersionDatabase::getMainThreadBusyness() {
+		Reference<ClientInfo> keptClient = itr->second;
+		Reference<ClientInfo> discardedClient = client;
+		if (client->canReplace(itr->second)) {
+			std::swap(keptClient, discardedClient);
```
does this have an effect?

sfc-gh-ajbeamon replied:
My memory of writing this is a bit hazy, but it looks like what this is saying is that if you add a new client with the same protocol as one we've already added, then it will by default keep the one you've already added. However, if the new one is in some way preferable (as determined by `canReplace`), then it will use the new one instead.

sfc-gh-anoyes reviewed (approved):
LGTM. As one final assurance I built this PR using TSAN, and a TSAN enabled 6.3 libfdb_c to use as an external client. TSAN didn't complain.
